crates/bindings-csharp/Codegen/Module.cs

```diff
@@ -294,36 +393,42 @@ public ReducerDeclaration(GeneratorAttributeSyntaxContext context)
         {
             throw new Exception($"Reducer {method} must return void");
         }
+        if (
+            method.Parameters.FirstOrDefault()?.Type is not INamedTypeSymbol namedType
+            || namedType.Name != "ReducerContext"
+        )
+        {
+            throw new Exception(
+                $"Reducer {method} must have a first argument of type ReducerContext"
+            );
+        }
 
         Name = method.Name;
         ExportName = attr.Name ?? Name;
         FullName = SymbolToName(method);
         Args = new(
-            method.Parameters.Select(p => new ReducerParamDeclaration(p)).ToImmutableArray()
+            method
+                .Parameters.Skip(1)
+                .Select(p => new MemberDeclaration(p.Name, p.Type))
+                .ToImmutableArray()
         );
         Scope = new Scope(methodSyntax.Parent as MemberDeclarationSyntax);
     }
 
-    private IEnumerable<ReducerParamDeclaration> NonContextArgs => Args.Where(a => !a.IsContextArg);
-
     public KeyValuePair<string, string> GenerateClass()
     {
+        var args = string.Join(", ", Args.Select(a => $"{a.Name}.Read(reader)").Prepend("ctx"));
         var class_ = $$"""
             class {{Name}}: SpacetimeDB.Internal.IReducer {
-                {{MemberDeclaration.GenerateBsatnFields(Accessibility.Private, NonContextArgs)}}
+                {{MemberDeclaration.GenerateBsatnFields(Accessibility.Private, Args)}}
 
                 public SpacetimeDB.Internal.ReducerDef MakeReducerDef(SpacetimeDB.BSATN.ITypeRegistrar registrar) => new (
                     "{{ExportName}}",
-                    [{{MemberDeclaration.GenerateDefs(NonContextArgs)}}]
+                    [{{MemberDeclaration.GenerateDefs(Args)}}]
                 );
 
-                public void Invoke(BinaryReader reader, SpacetimeDB.ReducerContext ctx) {
-                    {{FullName}}(
-                        {{string.Join(
-                            ", ",
-                            Args.Select(a => a.IsContextArg ? "ctx" : $"{a.Name}.Read(reader)")
-                        )}}
-                    );
+                public void Invoke(BinaryReader reader, SpacetimeDB.Internal.IReducerContext ctx) {
+                    {{FullName}}((SpacetimeDB.ReducerContext){{args}});
```

Comment:
Nit: I'd move `(SpacetimeDB.ReducerContext)` into the string in the `Prepend` above, it took me a moment to understand that it's not casting all the args, but that the first one happens to be our context, since they're declared and used in different places.